datadogfirehosenozzle/datadog_firehose_nozzle.go

```diff
@@ -84,20 +113,56 @@ func (d *DatadogFirehoseNozzle) postToDatadog() error {
 		select {
 		case <-ticker.C:
 			d.postMetrics()
+		case err := <-d.errs:
+			d.handleError(err)
+			return err
 		case envelope := <-d.messages:
 			if !d.keepMessage(envelope) {
 				continue
 			}
+			d.handleMessage(envelope)
+			d.client.ProcessMetric(envelope)
+		case <-d.stopper:
+			return nil
+		}
+	}
+}
 
+func (d *DatadogFirehoseNozzle) work() {
+	for {
+		select {
+		case envelope := <-d.messages:
+			if !d.keepMessage(envelope) {
+				continue
+			}
 			d.handleMessage(envelope)
-			d.client.AddMetric(envelope)
-		case err := <-d.errs:
-			d.handleError(err)
-			return err
+			d.client.ProcessMetric(envelope)
+		case <-d.stopWorkers:
+			return
 		}
 	}
 }
 
+func (d *DatadogFirehoseNozzle) startWorkers() {
+	for i := 0; i < d.config.NumWorkers; i++ {
+		go d.work()
+	}
+}
+
+func (d *DatadogFirehoseNozzle) stop() {
+	go func() {
+		for i := 0; i < d.config.NumWorkers; i++ {
+			d.stopWorkers <- true
+		}
+	}()
+}
+
+func (d *DatadogFirehoseNozzle) Stop() {
+	go func() {
+		d.stopper <- true
+	}()
+}
+
 func (d *DatadogFirehoseNozzle) postMetrics() {
 	err := d.client.PostMetrics()
 	if err != nil {
```

Comment:
Is an error here really a `Fatal` error? It feels like it just means we failed to submit some metrics.

Reply:
It is not, I'm gonna fix that